smolyay/samples.py

```diff
@@ -0,0 +1,484 @@
+            points = -numpy.cos(
+                numpy.pi * numpy.linspace(0, self.degree, self.degree + 1) / self.degree
+            )
+        else:
+            # special case where degree == 0
+            points = numpy.zeros(1)
+        self._points = self._scale_to_domain(points, [-1, 1])
+
+
+class NestedClenshawCurtisPointSet(NestedUnidimensionalPointSet):
+    r"""Generate nested Clenshaw Curtis points
+
+    The :attr:`points` for this interpolation scheme are the nested Clenshaw
+    Curtis points, which are the extrema of the Chebyshev polynomials of the
+    first kind with degree :math:`n` where n is part of a sequence
+     
+    .. math::
+
+        n = \begin{cases}
+            0 & \text{ if } L = 0\\ 
+            2^{L} & \text{ if } L > 0 
+        \end{cases}
+
+    where L is a whole number.
+
+    As the extrema of the Chebyshev polynomials of degree :math:`n` are nested,
+    meaning any :math:`n` will have extrema at the same points as every n that
+    precedes it in the sequence. The extrema are organized into levels such that
+    a level L will contain the extrema of Chebyshev polynomial n(L) that are
+    not the extrema of any Chebyshev polynomial n(k) where k is a whole number
+    and k < L.
+
+    The order :math:`o(L) of the nested Clenshaw Curtis set, which describes the
+    number of cummulative points at each level, is used to determine the 
+    number of points at each individual level. The order of the Clenshaw Curtis
+    set is described by the following equation:
+
+    .. math::
+
+        o(L) = \begin{cases}
+                1 & \text{ if } L = 0\\ 
+                2^{L} + 1 & \text{ if } L > 0 
+        \end{cases}
+
+    which leads to a sequence :math:`{1, 3, 5, 9, 17, ...}`.
+
+    Determining the number of points each level is then
+
+    .. math::
+        num_per_level(L) = o(L) - o(L - 1)
+
+    The points are then scaled from the domain :math:`[-1, 1]` to the domain
+    specified by the :attr:`domain`.
+
+    Parameters
+    ----------
+    domain: list
+        Domain of the sample points.
+
+    max_level: int
+        The maximum level the points are used for.
+    """
+
+    def _create(self):
+        r"""Create the points in the set.
+
+        Generating nested extrema of chebyshev polynomials of the first kind.
+        """
+        # create properties for levels, level 0 is a special case with 1 point
+        rule = lambda x: 1 if x == 0 else 2**x + 1
+        self._num_per_level = numpy.ones(self.max_level + 1, dtype=int)
+        self._num_per_level[1:] = [
+            rule(i) - rule((i - 1)) for i in range(1, self.max_level + 1)
+        ]
+        self._end_level = numpy.cumsum(self._num_per_level)
+        self._start_level = self._end_level - self._num_per_level
+
+        # points, level 0 is a special case only 0 as a point
+        num_points = self._end_level[-1]
+        points = numpy.zeros(num_points, dtype=float)
+        for i in range(1, self.max_level + 1):
```

Comment:
One thing that I'm noticing looking at all of these is that we are defining `max_level` as the inclusive upper bound, even though we are starting from level 0. That means that there are `max_level + 1` levels. Is this the behavior we want? Our usual conventions of levels start from level 1, meaning that there are `max_level` levels.

It could potentially make more sense to have a `num_levels` instead, as this unambiguously then runs from `0, ..., num_levels-1`.

Reply:
`max_level` is equivalent to the level of exactness that we've been using in all the documentation and papers, so changing it could be more confusing. Surrogates with only level 0 have one sampling point, and I would much prefer to have 0 be the trivial case with a single point than for 1 to be the trivial case.

Reply:
> `max_level` is equivalent to the level of exactness that we've been using in all the documentation and papers, so changing it could be more confusing.

I think we had discussed with Dr. K, and I was talking with Mohammadreza today, about how for the general case of mixed basis functions / anisotropic sampling there isn't as strong of a concept of level of exactness per dimension.

> Surrogates with only level 0 have one sampling point, and I would much prefer to have 0 be the trivial case with a single point than for 1 to be the trivial case.

Why?  FWIW, our paper has ks that are 1-indexed, so 0 is never allowed as a level. We have made them be 0-indexed in Python because that's the convention for this programming language, and so I find it pretty confusing to have the upper bound be inclusive. It's an anti-pattern for 0-indexed programming languages. We could still use `max_level` but make it exclusive, if you like that better. It's functionally the same as naming it `num_levels`, but I think `num_levels` is less ambiguous.

> Doing it by `num_levels` would also mean there is no behavior for an input of 0, or if there is behavior, it would be to return an empty array.

It probably doesn't make sense from a practical point of view to construct a set with `num_levels = 0`, but it would be an empty set if it were allowed.